File: src/api/multipass_discovery.cpp
/*
 * Copyright (C) Elemento.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; version 3.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 */

#include "multipass_discovery.h"

#include <multipass/format.h>
#include <multipass/logging/log.h>
#include <multipass/utils.h>

#include <QDir>
#include <QFileInfo>

#include <cstdlib>

namespace mp = multipass;
namespace mpl = multipass::logging;

namespace
{
constexpr auto category = "multipass-discovery";

bool file_exists(const QString& path)
{
    return QFileInfo::exists(path);
}

std::string read_file(const QString& path)
{
    return mp::utils::contents_of(path);
}

QString multipass_root_cert_path()
{
#if defined(MULTIPASS_PLATFORM_WINDOWS)
    const auto program_data = qEnvironmentVariable("PROGRAMDATA", "C:\\ProgramData");
    return QDir(program_data).filePath("Multipass/data/multipassd/multipass_root_cert.pem");
#elif defined(MULTIPASS_PLATFORM_APPLE)
    return QStringLiteral("/usr/local/etc/multipassd/multipass_root_cert.pem");
#else
    const auto snap_common = qEnvironmentVariable("SNAP_COMMON");
    if (!snap_common.isEmpty())
    {
        const auto snap_path =
            QDir(snap_common).filePath("data/multipassd/multipass_root_cert.pem");
        if (file_exists(snap_path))
            return snap_path;
    }
    const auto snap_host =
        QStringLiteral("/var/snap/multipass/common/data/multipassd/multipass_root_cert.pem");
    if (file_exists(snap_host))
        return snap_host;
    return QStringLiteral("/usr/local/etc/multipassd/multipass_root_cert.pem");
#endif
}

QString multipass_client_cert_dir()
{
#if defined(MULTIPASS_PLATFORM_WINDOWS)
    const auto appdata = qEnvironmentVariable("APPDATA");
    return QDir(appdata).filePath("multipass-client-certificate");
#elif defined(MULTIPASS_PLATFORM_APPLE)
    const auto home = qEnvironmentVariable("HOME");
    return QDir(home).filePath("Library/Application Support/multipass-client-certificate");
#else
    const auto xdg = qEnvironmentVariable("XDG_DATA_HOME");
    if (!xdg.isEmpty())
        return QDir(xdg).filePath("multipass-client-certificate");
    const auto home = qEnvironmentVariable("HOME");
    return QDir(home).filePath(".local/share/multipass-client-certificate");
#endif
}

bool address_looks_available(const std::string& address)
{
    constexpr std::string_view unix_prefix = "unix:";
    if (address.rfind(unix_prefix, 0) == 0)
        return file_exists(QString::fromStdString(address.substr(unix_prefix.size())));
    // TCP: allow the connection attempt.
    return true;
}
} // namespace

std::string mp::api::default_multipass_server_address()
{
#if defined(MULTIPASS_PLATFORM_WINDOWS)
    return "localhost:50051";
#elif defined(MULTIPASS_PLATFORM_APPLE)
    return "unix:/var/run/multipass_socket";
#else
    auto unix_if_exists = [](const QString& socket_path) -> std::string {
        if (file_exists(socket_path))
            return fmt::format("unix:{}", socket_path.toStdString());
        return {};
    };

    const auto snap_common = qEnvironmentVariable("SNAP_COMMON");
    if (!snap_common.isEmpty())
    {
        if (auto addr = unix_if_exists(QDir(snap_common).filePath("multipass_socket"));
            !addr.empty())
            return addr;
    }
    if (auto addr =
            unix_if_exists(QStringLiteral("/var/snap/multipass/common/multipass_socket"));
        !addr.empty())
        return addr;
    return "unix:/run/multipass_socket";
#endif
}

std::optional<mp::api::MultipassConnection> mp::api::discover_multipass_connection(
    const std::string& address_override)
{
    try
    {
        MultipassConnection conn;
        conn.address =
            address_override.empty() ? default_multipass_server_address() : address_override;

        if (!address_looks_available(conn.address))
        {
            mpl::log_message(mpl::Level::warning,
                             category,
                             fmt::format("Multipass socket/address not available: {}",
                                         conn.address));
            return std::nullopt;
        }

        const auto root_path = multipass_root_cert_path();
        if (!file_exists(root_path))
        {
            mpl::log_message(
                mpl::Level::warning,
                category,
                fmt::format("Multipass root CA missing: {}", root_path.toStdString()));
            return std::nullopt;
        }

        const auto cert_dir = multipass_client_cert_dir();
        const auto cert_path = QDir(cert_dir).filePath("multipass_cert.pem");
        const auto key_path = QDir(cert_dir).filePath("multipass_cert_key.pem");
        if (!file_exists(cert_path) || !file_exists(key_path))
        {
            mpl::log_message(mpl::Level::warning,
                             category,
                             fmt::format("Multipass client certificates missing under {}",
                                         cert_dir.toStdString()));
            return std::nullopt;
        }

        conn.root_cert_pem = read_file(root_path);
        conn.client_cert_pem = read_file(cert_path);
        conn.client_key_pem = read_file(key_path);

        mpl::log_message(
            mpl::Level::info,
            category,
            fmt::format("Discovered Multipass at {} (certs under {})",
                        conn.address,
                        cert_dir.toStdString()));
        return conn;
    }
    catch (const std::exception& e)
    {
        mpl::log_message(mpl::Level::warning,
                         category,
                         fmt::format("Failed to discover Multipass connection: {}", e.what()));
        return std::nullopt;
    }
}

std::shared_ptr<grpc::Channel> mp::api::make_channel_with_pems(const std::string& server_address,
                                                               const std::string& root_cert_pem,
                                                               const std::string& client_cert_pem,
                                                               const std::string& client_key_pem)
{
    grpc::SslCredentialsOptions opts;
    opts.pem_root_certs = root_cert_pem;
    opts.pem_cert_chain = client_cert_pem;
    opts.pem_private_key = client_key_pem;

    grpc::ChannelArguments channel_args;
    channel_args.SetString(GRPC_ARG_DEFAULT_AUTHORITY, "localhost");
    return grpc::CreateCustomChannel(server_address,
                                     grpc::SslCredentials(opts),
                                     channel_args);
}

File: src/client/gui/lib/multipass_discovery.dart
import 'dart:io';

import 'logger.dart';
import 'platform/platform.dart';

/// Credentials and address needed to talk to a stock Multipass daemon.
class MultipassConnectionConfig {
  final Uri address;
  final List<int> rootCert;
  final List<int> clientCert;
  final List<int> clientKey;

  const MultipassConnectionConfig({
    required this.address,
    required this.rootCert,
    required this.clientCert,
    required this.clientKey,
  });
}

/// Discovers stock Multipass daemon endpoints and TLS material on disk.
///
/// Returns null when Multipass does not appear to be installed / reachable
/// (missing socket, root CA, or client certificates).
MultipassConnectionConfig? discoverMultipassConnection() {
  try {
    final address = _multipassServerAddress();
    if (!_addressLooksAvailable(address)) {
      logger.w('Multipass socket/address not available: $address');
      return null;
    }

    final rootCertPath = _multipassRootCertPath();
    final rootFile = File(rootCertPath);
    if (!rootFile.existsSync()) {
      logger.w('Multipass root CA missing: $rootCertPath');
      return null;
    }

    final certDir = _multipassClientCertDir();
    final certFile = File('$certDir/multipass_cert.pem');
    final keyFile = File('$certDir/multipass_cert_key.pem');
    if (!certFile.existsSync() || !keyFile.existsSync()) {
      logger.w('Multipass client certificates missing under $certDir');
      return null;
    }

    logger.i('Discovered Multipass at $address (certs under $certDir)');
    return MultipassConnectionConfig(
      address: address,
      rootCert: rootFile.readAsBytesSync(),
      clientCert: certFile.readAsBytesSync(),
      clientKey: keyFile.readAsBytesSync(),
    );
  } catch (e, st) {
    logger.w('Failed to discover Multipass connection', error: e, stackTrace: st);
    return null;
  }
}

Uri _multipassServerAddress() {
  final override = Platform.environment['ELP_MULTIPASS_ADDRESS'];
  if (override != null && override.isNotEmpty) {
    return _parseServerAddress(override);
  }

  if (Platform.isMacOS) {
    return _parseServerAddress('unix:/var/run/multipass_socket');
  }
  if (Platform.isWindows) {
    return _parseServerAddress('localhost:50051');
  }
  // Linux: prefer snap common path, then classic /run.
  final snapCommon = Platform.environment['SNAP_COMMON'];
  if (snapCommon != null && snapCommon.isNotEmpty) {
    final snapSocket = File('$snapCommon/multipass_socket');
    if (snapSocket.existsSync()) {
      return _parseServerAddress('unix:$snapCommon/multipass_socket');
    }
  }
  const snapHostPath = '/var/snap/multipass/common/multipass_socket';
  if (File(snapHostPath).existsSync()) {
    return _parseServerAddress('unix:$snapHostPath');
  }
  return _parseServerAddress('unix:/run/multipass_socket');
}

String _multipassRootCertPath() {
  if (Platform.isMacOS) {
    return '/usr/local/etc/multipassd/multipass_root_cert.pem';
  }
  if (Platform.isWindows) {
    final programData =
        Platform.environment['PROGRAMDATA'] ?? r'C:\ProgramData';
    return '$programData\\Multipass\\data\\multipassd\\multipass_root_cert.pem';
  }
  final snapCommon = Platform.environment['SNAP_COMMON'];
  if (snapCommon != null && snapCommon.isNotEmpty) {
    final snapPath =
        '$snapCommon/data/multipassd/multipass_root_cert.pem';
    if (File(snapPath).existsSync()) return snapPath;
  }
  const snapHostCert =
      '/var/snap/multipass/common/data/multipassd/multipass_root_cert.pem';
  if (File(snapHostCert).existsSync()) return snapHostCert;
  return '/usr/local/etc/multipassd/multipass_root_cert.pem';
}

String _multipassClientCertDir() {
  if (Platform.isMacOS) {
    final home = mpPlatform.homeDirectory ??
        Platform.environment['HOME'] ??
        '';
    return '$home/Library/Application Support/multipass-client-certificate';
  }
  if (Platform.isWindows) {
    final appData = Platform.environment['APPDATA'] ?? '';
    return '$appData\\multipass-client-certificate';
  }
  // Linux XDG data home
  final xdg = Platform.environment['XDG_DATA_HOME'];
  if (xdg != null && xdg.isNotEmpty) {
    return '$xdg/multipass-client-certificate';
  }
  final home = mpPlatform.homeDirectory ??
      Platform.environment['HOME'] ??
      '';
  return '$home/.local/share/multipass-client-certificate';
}

Uri _parseServerAddress(String address) {
  final unixRegex = RegExp('unix:(.+)');
  final unixSocketPath = unixRegex.firstMatch(address)?.group(1);
  if (unixSocketPath != null) {
    return Uri(scheme: InternetAddressType.unix.name, path: unixSocketPath);
  }
  final tcpRegex = RegExp(r'^(.+):(\d+)$');
  final tcpMatch = tcpRegex.firstMatch(address);
  if (tcpMatch != null) {
    return Uri(host: tcpMatch.group(1), port: int.parse(tcpMatch.group(2)!));
  }
  throw FormatException('Invalid Multipass server address: $address');
}

bool _addressLooksAvailable(Uri address) {
  if (address.scheme == InternetAddressType.unix.name) {
    return File(address.path).existsSync();
  }
  // TCP: cannot cheaply probe without connecting; allow attempt when configured.
  return true;
}
